presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,242 @@
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class RandomizeNullKeyInOuterJoin
+        implements PlanOptimizer
+{
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public RandomizeNullKeyInOuterJoin(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (getJoinDistributionType(session).canPartition() && randomizeOuterJoinNullKeyEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, functionAndTypeManager, idAllocator, variableAllocator, new HashMap<>()), plan, new Context());
+        }
+
+        return plan;
+    }
+
+    private static class Context
+    {
+        private final Stack<PlanNode> parentNodes = new Stack<>();
+
+        public PlanNode getParentNode()
+        {
+            return parentNodes.peek();
+        }
+
+        public void setParentNode(PlanNode parentNode)
+        {
+            parentNodes.push(parentNode);
+        }
+
+        public void popParentNode()
+        {
+            parentNodes.pop();
+        }
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Context>
+    {
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final PlanVariableAllocator planVariableAllocator;
+        private final Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap;
+
+        private Rewriter(Session session,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, PlanVariableAllocator planVariableAllocator,
+                Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.keyToRandomKeyMap = requireNonNull(keyToRandomKeyMap, "keyToRandomKeyMap is null");
```

Comment:
Why is this passed in? I thought we use only a single instance for all? I say just initialize it in the class. No need to pass it in. If you do need to pass it in, put it in context.

Reply:
Update the code to initialize the map in class constructor.